### PR TITLE
Master localize number menu preview lul

### PR DESCRIPTION
**Pull request by LucasLefevre**
## Description:

description of this task, what is implemented and why it is implemented that way.

Odoo task ID : [TASK_ID](https://www.odoo.com/web#id=TASK_ID&action=333&active_id=2328&model=project.task&view_type=form&cids=1&menu_id=4720)

## review checklist

- [ ] feature is organized in plugin, or UI components
- [ ] support of duplicate sheet (deep copy)
- [ ] in model/core: ranges are Range object, and can be adapted (adaptRanges)
- [ ] in model/UI: ranges are strings (to show the user)
- [ ] undo-able commands (uses this.history.update)
- [ ] multiuser-able commands (has inverse commands and transformations where needed)
- [ ] new/updated/removed commands are documented
- [ ] exportable in excel
- [ ] translations (\_lt("qmsdf %s", abc))
- [ ] unit tested
- [ ] clean commented code
- [ ] track breaking changes
- [ ] doc is rebuild (npm run doc)
- [ ] status is correct in Odoo

### CODE REVIEW

File: tests/menu_items_registry.test.ts [hunk 1]
```diff
@@ -874,6 +876,34 @@ describe("Menu Item actions", () => {
...
+      ["format_number_currency", "$1 000,12"],
+      ["format_number_currency_rounded", "$1 000"],
+      ["format_number_date", "26/09/2023"],
+      ["format_number_time", "22:43:00"],
+      ["format_number_date_time", "26/09/2023 22:43:00"],
+      ["format_number_duration", "27:51:38"],
+    ])("number formatting description with custom locale", (actionId, expectedDescription) => {
+      updateLocale(model, FR_LOCALE);
+      const action = getNode(["format", "format_number", actionId]);
+      expect(action.description?.(env)).toBe(expectedDescription);
```
Comment (hokolomopo) [inline tests/menu_items_registry.test.ts:904]: useless `?.`

File: tests/menu_items_registry.test.ts [hunk 2]
```diff
@@ -874,6 +876,34 @@ describe("Menu Item actions", () => {
...
+      ["format_number_currency_rounded", "$1,000"],
+      ["format_number_date", "9/26/2023"],
+      ["format_number_time", "10:43:00 PM"],
+      ["format_number_date_time", "9/26/2023 10:43:00 PM"],
+      ["format_number_duration", "27:51:38"],
+    ])("number formatting description with default locale", (actionId, expectedDescription) => {
+      const action = getNode(["format", "format_number", actionId]);
+      expect(action.description?.(env)).toBe(expectedDescription);
+    });
+    test.each([
```
Comment (hokolomopo) [inline tests/menu_items_registry.test.ts:892]: nitpick: newline between tests is prettier 🙃

File: src/model.ts [hunk 3]
```diff
@@ -387,6 +389,7 @@ export class Model extends EventBus<any> implements CommandDispatcher {
       ...config,
       mode: config.mode || "normal",
       custom: config.custom || {},
+      defaultCurrencyFormat: config.defaultCurrencyFormat || "[$$]#,##0.00",
```
Comment (hokolomopo) [inline src/model.ts:392]: We could also check if the given format is valid.

File: src/helpers/format.ts [hunk 4]
```diff
@@ -527,6 +527,42 @@ export function detectNumberFormat(content: string): Format | undefined {
...
+  const decimalRepresentation = decimalPlaces ? "." + "0".repeat(decimalPlaces) : "";
+  const numberFormat = "#,##0" + decimalRepresentation;
+  let textExpression = `${code} ${symbol}`.trim();
+  if (position === "after" && code) {
+    textExpression = " " + textExpression;
+  }
+  return insertText(textExpression, position, numberFormat);
+}
+
+function insertText(text: string, position: "before" | "after", format: Format): Format {
```
Comment (hokolomopo) [inline src/helpers/format.ts:544]: `insertTextInFormat` maybe ?

File: tests/helpers/format.test.ts [hunk 5]
```diff
@@ -895,3 +900,40 @@ describe("formatValue on date and time", () => {
...
+  test("round format", () => {
+    expect(roundFormat("#,##0.00")).toBe("#,##0");
+    expect(roundFormat("#,##0")).toBe("#,##0");
+    expect(roundFormat("#,##0.")).toBe("#,##0");
+    expect(roundFormat("0.00%")).toBe("0%");
+    expect(roundFormat("[$$]#,##0.00")).toBe("[$$]#,##0");
+    expect(roundFormat("ddd-mm-yyyy")).toBe("ddd-mm-yyyy");
+  });
+});
+describe("create currency format", () => {
```
Comment (hokolomopo) [inline tests/helpers/format.test.ts:914]:
```suggestion
});

describe("create currency format", () => {
```

File: tests/menu_items_registry.test.ts [hunk 6]
```diff
@@ -883,22 +913,51 @@ describe("Menu Item actions", () => {
...
-    test("Currency", () => {
-      doAction(["format", "format_number", "format_number_currency"], env);
-      expect(dispatch).toHaveBeenCalledWith("SET_FORMATTING", {
-        sheetId: env.model.getters.getActiveSheetId(),
-        target: env.model.getters.getSelectedZones(),
-        format: "[$$]#,##0.00",
-      });
+    test("currency format with default currency", () => {
+      const action = getNode(["format", "format_number", "format_number_currency"]);
+      expect(action.description?.(env)).toBe("$1,000.12");
```
Comment (hokolomopo) [inline tests/menu_items_registry.test.ts:918]:
```suggestion
      expect(action.description(env)).toBe("$1,000.12");
```

Comment (hokolomopo) [reply tests/menu_items_registry.test.ts:918]: Same with all the other `action.description?.()`

File: tests/menu_items_registry.test.ts [hunk 7]
```diff
@@ -883,22 +913,51 @@ describe("Menu Item actions", () => {
...
+    test("rounded currency format with custom currency", () => {
+      const model = new Model({}, { defaultCurrencyFormat: "[$€]#,##0.000" });
+      env = makeTestEnv({ model });
+      const action = getNode(["format", "format_number", "format_number_currency_rounded"]);
+      expect(action.description?.(env)).toBe("€1,000");
+      action.execute?.(env);
+      expect(getCell(model, "A1")?.format).toBe("[$€]#,##0");
+    });
+
+    test("rounded currency format is invisible if the default format is already rounded", () => {
```
Comment (hokolomopo) [inline tests/menu_items_registry.test.ts:948]: nitpick: should be `if the custom format...` to be coherent with other test name. Maybe change eveything to `custom default currency` ?

**Review by hokolomopo — COMMENTED** [review]
Mostly details/nitpicks 👍 
The runbot is red AF tho

File: tests/helpers/format.test.ts [hunk 8]
```diff
@@ -895,3 +900,40 @@ describe("formatValue on date and time", () => {
...
+  const code = "O$OO";
+  const currency: Currency = {
+    code,
+    symbol,
+    decimalPlaces: 2,
+    name: "Odoo Dollar",
+    position: "before",
+  };
+  test("full currency", () => {
+    expect(createCurrencyFormat(currency)).toBe("[$O$OO θ]#,##0.00");
```
[the remark this answers is not in the document]
Comment (LucasLefevre) [reply tests/helpers/format.test.ts:925]: Github Copilot is my best friend :rocket:

File: src/model.ts (hunk 3 above)
Comment (laa-odoo) [reply src/model.ts:392]: indeed yes

**Review by laa-odoo — APPROVED** [review]

**Comment by laa-odoo** [comment]
@robodoo r+

**Comment by LucasLefevre** [comment]
@laa-odoo le classique :smile: 
robodoo rebase-ff

**Pull request closed without merging**